benchmark/rocblas/blas1/asum.cpp

```diff
@@ -0,0 +1,145 @@
+  str << size;
+  return str.str();
+}
+
+template <typename scalar_t, typename... args_t>
+static inline void rocblas_asum_f(args_t&&... args) {
+  if constexpr (std::is_same_v<scalar_t, float>) {
+    CHECK_ROCBLAS_STATUS(rocblas_sasum(std::forward<args_t>(args)...));
+  } else if constexpr (std::is_same_v<scalar_t, double>) {
+    CHECK_ROCBLAS_STATUS(rocblas_dasum(std::forward<args_t>(args)...));
+  }
+  return;
+}
+
+template <typename scalar_t>
+void run(benchmark::State& state, rocblas_handle& rb_handle, index_t size,
+         bool* success) {
+  // Google-benchmark counters are double.
+  double size_d = static_cast<double>(size);
+  state.counters["size"] = size_d;
+  state.counters["n_fl_ops"] = 2 * size_d;
+  state.counters["bytes_processed"] = size_d * sizeof(scalar_t);
+
+  using data_t = scalar_t;
+
+  // Create data
+  std::vector<data_t> v1 = blas_benchmark::utils::random_data<data_t>(size);
+  scalar_t vr;
+
+  {
+    // Device memory allocation & H2D Copy
+    blas_benchmark::utils::HIPVector<scalar_t> d_v1(size, v1.data());
+    blas_benchmark::utils::HIPScalar<scalar_t> d_vr(vr);
+
+#ifdef BLAS_VERIFY_BENCHMARK
+    // Reference asum
+    data_t vr_ref = reference_blas::asum(size, v1.data(), 1);
+
+    // Rocblas verification asum
+    data_t vr_temp = 0;
+
+    {
+      // Temp result on device (copied back to Host upon destruction)
+      blas_benchmark::utils::HIPScalar<scalar_t, true> d_vr_temp(vr_temp);
+      // rocBLAS function call
+      rocblas_asum_f<scalar_t>(rb_handle, size, d_v1, 1, d_vr_temp);
+    }
+
+    if (!utils::almost_equal(vr_temp, vr_ref)) {
+      std::ostringstream err_stream;
+      err_stream << "Value mismatch: " << vr_temp << "; expected " << vr_ref;
+      const std::string& err_str = err_stream.str();
+      state.SkipWithError(err_str.c_str());
+      *success = false;
+    };
+#endif
+
+    auto blas_warmup = [&]() -> void {
+      rocblas_asum_f<scalar_t>(rb_handle, size, d_v1, 1, d_vr);
+      CHECK_HIP_ERROR(hipStreamSynchronize(NULL));
+      return;
+    };
+
+    hipEvent_t start, stop;
+    CHECK_HIP_ERROR(hipEventCreate(&start));
+    CHECK_HIP_ERROR(hipEventCreate(&stop));
+
+    auto blas_method_def = [&]() -> std::vector<hipEvent_t> {
+      CHECK_HIP_ERROR(hipEventRecord(start, NULL));
+      rocblas_asum_f<scalar_t>(rb_handle, size, d_v1, 1, d_vr);
+      CHECK_HIP_ERROR(hipEventRecord(stop, NULL));
+      CHECK_HIP_ERROR(hipEventSynchronize(stop));
+      return std::vector{start, stop};
+    };
+
+    // Warm up to avoid benchmarking data transfer
+    blas_benchmark::utils::warmup(blas_warmup);
+
+    blas_benchmark::utils::init_counters(state);
+
+    // Measure
+    for (auto _ : state) {
+      std::tuple<double, double> times =
+          blas_benchmark::utils::timef_hip(blas_method_def);
+
+      // Report
+      blas_benchmark::utils::update_counters(state, times);
+    }
+
+    blas_benchmark::utils::calc_avg_counters(state);
```

Comment:
I see you usually don't destroy the `hipEvent_t` is it ok not to do it?

Reply:
Well spotted, it should be destroyed indeed!